internal/provider/data_source.go

```diff
@@ -177,6 +178,10 @@ The program must also be executable according to the platform where Terraform is
 	workingDir := config.WorkingDir.ValueString()
 
 	cmd := exec.CommandContext(ctx, filteredProgram[0], filteredProgram[1:]...)
+	if errors.Is(cmd.Err, exec.ErrDot) {
```

Comment:
Hacky idea incoming, maybe it'll spur out a better idea 😆 :

- Instead of `go install`-ing the test program, you could `go build` the test program into the working directory
- Then update the `PATH` variable to include the working directory `PATH=./` with `t.Setenv`
- Execute the provider test with just the binary name `tf-acc-external-data-source`

Reply:
I believe to trigger the `ErrDot`, you want the resulting path that finds the executable to be a relative path, that snippet looks like the resulting path would be absolute from `os.Getwd()`. Try something like this:

``` go
func TestDataSource_CurrentDir(t *testing.T) {
	p := os.Getenv("PATH")
	t.Setenv("PATH", fmt.Sprintf("%s:./", p))

	resource.UnitTest(t, resource.TestCase{
		ProtoV5ProviderFactories: protoV5ProviderFactories(),
		Steps: []resource.TestStep{
			{
				Config: fmt.Sprintf(`
					data "external" "test" {
						program = [%[1]q]
				
						query = {
							value = null,
							value2 = ""
						}
					}
				`, "tf-acc-external-data-source"),
				Check: resource.ComposeAggregateTestCheckFunc(
					resource.TestCheckResourceAttr("data.external.test", "result.value", ""),
					resource.TestCheckResourceAttr("data.external.test", "result.value2", ""),
				),
			},
		},
	})
}

```

Reply:
Prior to adding workaround code:

```go
	_, err = exec.LookPath(filteredProgram[0])
	if errors.Is(err, exec.ErrDot) {
		err = nil
	}

	/* ... */

	cmd := exec.CommandContext(ctx, filteredProgram[0], filteredProgram[1:]...)
	if errors.Is(cmd.Err, exec.ErrDot) {
		cmd.Err = nil
	}
```

`TestDataSource_CurrentDir` test fails with:

```
        Error: External Program Lookup Failed
        
          with data.external.test,
          on terraform_plugin_test.tf line 3, in data "external" "test":
           3: 						program = ["tf-acc-external-data-source"]
        
        The data source received an unexpected error while attempting to parse the
        query. The data source received an unexpected error while attempting to find
        the program.
        
        The program must be accessible according to the platform where Terraform is
        running.
        
        If the expected program should be automatically found on the platform where
        Terraform is running, ensure that the program is in an expected directory. On
        Unix-based platforms, these directories are typically searched based on the
        '$PATH' environment variable. On Windows-based platforms, these directories
        are typically searched based on the '%PATH%' environment variable.
        
        If the expected program is relative to the Terraform configuration, it is
        recommended that the program name includes the interpolated value of
        'path.module' before the program name to ensure that it is compatible with
        varying module usage. For example: "${path.module}/my-program"
        
        The program must also be executable according to the platform where Terraform
        is running. On Unix-based platforms, the file on the filesystem must have the
        executable bit set. On Windows-based platforms, no action is typically
        necessary.
        
        Platform: darwin
        Program: "tf-acc-external-data-source"
        Error: exec: "tf-acc-external-data-source": cannot run executable found
        relative to current directory
```